File: Preferences/Preferences/RubberbandEngineTransformer.m
//
//  EngineTransformer.m
//  Preferences
//
//  Created by Christopher Snowhill on 2/11/25.
//

#import <Foundation/Foundation.h>

#import "RubberbandEngineTransformer.h"

@implementation RubberbandEngineR3Transformer
+ (Class)transformedValueClass {
	return [NSNumber class];
}
+ (BOOL)allowsReverseTransformation {
	return NO;
}

- (id)transformedValue:(id)value {
	if(value == nil) return @(YES);

	if([value isKindOfClass:[NSString class]]) {
		NSString *stringValue = value;
		if([stringValue isEqualToString:@"disabled"] ||
		   [stringValue isEqualToString:@"finer"]) {
			return @(NO);
		}
	}

	return @(YES);
}

@end

@implementation RubberbandEngineEnabledTransformer
+ (Class)transformedValueClass {
	return [NSNumber class];
}
+ (BOOL)allowsReverseTransformation {
	return NO;
}

- (id)transformedValue:(id)value {
	if(value == nil) return @(YES);

	if([value isKindOfClass:[NSString class]]) {
		NSString *stringValue = value;
		if([stringValue isEqualToString:@"disabled"]) {
			return @(NO);
		}
	}

	return @(YES);
}

@end

@implementation RubberbandEngineHiddenTransformer
+ (Class)transformedValueClass {
	return [NSNumber class];
}
+ (BOOL)allowsReverseTransformation {
	return NO;
}

- (id)transformedValue:(id)value {
	if(value == nil) return @(YES);

	if([value isKindOfClass:[NSString class]]) {
		NSString *stringValue = value;
		if([stringValue isEqualToString:@"disabled"]) {
			return @(NO);
		}
	}

	return @(YES);
}

@end

File: Transformers/RubberbandEngineTransformer.swift
import Foundation

@objc(RubberbandEngineR3Transformer)
class RubberbandEngineR3Transformer: ValueTransformer {
    override class func transformedValueClass() -> AnyClass {
        return NSNumber.self
    }
    
    override class func allowsReverseTransformation() -> Bool {
        return false
    }
    
    override func transformedValue(_ value: Any?) -> Any? {
        if value == nil {
            return NSNumber(value: true)
        }
        
        if let stringValue = value as? String {
            if stringValue == "disabled" || stringValue == "signalsmith" || stringValue == "finer" {
                return NSNumber(value: false)
            }
        }
        
        return NSNumber(value: true)
    }
}

@objc(RubberbandEngineEnabledTransformer)
class RubberbandEngineEnabledTransformer: ValueTransformer {
    override class func transformedValueClass() -> AnyClass {
        return NSNumber.self
    }
    
    override class func allowsReverseTransformation() -> Bool {
        return false
    }
    
    override func transformedValue(_ value: Any?) -> Any? {
        if value == nil {
            return NSNumber(value: true)
        }
        
        if let stringValue = value as? String {
            if stringValue == "disabled" || stringValue == "signalsmith" {
                return NSNumber(value: false)
            }
        }
        
        return NSNumber(value: true)
    }
}

@objc(RubberbandEngineHiddenTransformer)
class RubberbandEngineHiddenTransformer: ValueTransformer {
    override class func transformedValueClass() -> AnyClass {
        return NSNumber.self
    }
    
    override class func allowsReverseTransformation() -> Bool {
        return false
    }
    
    override func transformedValue(_ value: Any?) -> Any? {
        if value == nil {
            return NSNumber(value: true)
        }
        
        if let stringValue = value as? String {
            if stringValue == "disabled" {
                return NSNumber(value: false)
            }
        }
        
        return NSNumber(value: true)
    }
}
